visual_diff.py

```diff
@@ -27,54 +28,97 @@ def parse_args():
+      tuples for each token
+    """
     with open(filename) as f:
-        tokens = tokenize.generate_tokens(f.readline)
-        # Ignore non-significant whitespace
-        ignore_types = [token.NEWLINE, token.ENDMARKER, tokenize.NL,
-                        tokenize.COMMENT]
-        tokens = [tok for tok in tokens if tok.type not in ignore_types]
-        f.seek(0)
-        lines = [line.rstrip() for line in f.readlines()]
-
-    # We treat all constants as identical to other constants of the same type.
-    constant_types = (token.STRING, token.NUMBER)
+        contents = f.read()
+    toks = code_tokenize.tokenize(contents, lang=language)
+    toks = [t for t in toks if t.type not in ("newline", "comment")]
+    lines = [line.rstrip() for line in contents.split("\n")]
+    constant_types = ("string", "integer", "float", "indent", "dedent")
     token_array = numpy.array(
-            [tok.type if tok.type in constant_types else tok.string
-             for tok in tokens])
-    boundaries = [(tok.start, tok.end) for tok in tokens]
-    return token_array, lines, boundaries
+            [tok.type if tok.type in constant_types else tok.text
+             for tok in toks])
+
+    boundaries = []
+    for i, t in enumerate(toks):
+        try:
+            start = t.ast_node.start_point
+            end = t.ast_node.end_point
+            boundaries.append(((start[0] + 1, start[1]), (end[0] + 1, end[1])))
+        except AttributeError:
+            if t.type == "indent":
+                assert(t.new_line_before)
+                next_t = toks[i+1]
+                line = next_t.ast_node.start_point[0] + 1
+                boundaries.append(
+                    ((line, 0), (line, next_t.ast_node.start_point[1]-1)))
+            elif t.type == "dedent":
+                # We might be the very last token. Look backwards to the last
+                # non-dedent token to figure out what line we're on.
+                di = 0
+                prev_t = t
+                while prev_t.type == "dedent":
+                    di -= 1
+                    prev_t = toks[i + di]
+                line = prev_t.ast_node.end_point[0] + 2
+                # Eh... this is good enough, though not correct
+                boundaries.append(((line, 0), (line, 0)))
+            else:
+                print("UNEXPECTED TOKEN!", i, t, type(t), dir(t))
+                raise
+
+    return file_info.FileInfo(token_array, lines, boundaries)
+
+
+def get_text_width(args):
+    if args.gui_width is not None:
+        return args.gui_width
+    if args.language == "go":
+        return 100
+    return 80
 
 
 if __name__ == "__main__":
     args = parse_args()
-    data_a = get_tokens(args.filename_a)
-    data_b = get_tokens(args.filename_b or args.filename_a)
-    tokens_a = data_a[0]
-    tokens_b = data_b[0]
+    data_a = get_tokens(args.filename_a, args.language)
+    data_b = get_tokens(args.filename_b or args.filename_a, args.language)
```

Comment:
add a TODO for the future: it would be neat to compare across multiple languages.